In [65]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [66]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Q1

In [67]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/hmda_ca_c2.csv')

df.head()

year       age      race                  ethnicity        sex  \
0  2023  6. 65-74  3. Black  2. Not Hispanic or Latino  2. Female   
1  2023  4. 45-54  3. Black  2. Not Hispanic or Latino  2. Female   
2  2023  5. 55-64  5. White      1. Hispanic or Latino    1. Male   
3  2023  5. 55-64  5. White  2. Not Hispanic or Latino  2. Female   
4  2023  3. 35-44  2. Asian  2. Not Hispanic or Latino    1. Male   

  debt_to_income_ratio good_credit  income  
0           4. 40%-<50      1. Yes    79.0  
1           4. 40%-<50      1. Yes    77.0  
2           4. 40%-<50      1. Yes   720.0  
3           4. 40%-<50      1. Yes   590.0  
4           4. 40%-<50      1. Yes   285.0

## Q2


In [68]:
df['age']=df['age'].str.slice(3,)
df['race']=df['race'].str.slice(3,)
df['ethnicity']=df['ethnicity'].str.slice(3,)
df['sex']=df['sex'].str.slice(3,)
df['debt_to_income_ratio']=df['debt_to_income_ratio'].str.slice(3,)
df['good_credit']=df['good_credit'].str.slice(3,)

In [69]:
df.head()

year    age   race               ethnicity     sex debt_to_income_ratio  \
0  2023  65-74  Black  Not Hispanic or Latino  Female              40%-<50   
1  2023  45-54  Black  Not Hispanic or Latino  Female              40%-<50   
2  2023  55-64  White      Hispanic or Latino    Male              40%-<50   
3  2023  55-64  White  Not Hispanic or Latino  Female              40%-<50   
4  2023  35-44  Asian  Not Hispanic or Latino    Male              40%-<50   

  good_credit  income  
0         Yes    79.0  
1         Yes    77.0  
2         Yes   720.0  
3         Yes   590.0  
4         Yes   285.0

## Q3

In [70]:
np.percentile(df['income'], [10,70,99])


array([  61.  ,  177.  , 1216.15])

In [71]:

df['income_class']=df['income'].apply(lambda x: 'One Percent' if x>=1216.15 else 'Upper Class' if x>=177 else 'Middle Class' if x>=61 else "Low Income")


In [72]:
df.head()

year    age   race               ethnicity     sex debt_to_income_ratio  \
0  2023  65-74  Black  Not Hispanic or Latino  Female              40%-<50   
1  2023  45-54  Black  Not Hispanic or Latino  Female              40%-<50   
2  2023  55-64  White      Hispanic or Latino    Male              40%-<50   
3  2023  55-64  White  Not Hispanic or Latino  Female              40%-<50   
4  2023  35-44  Asian  Not Hispanic or Latino    Male              40%-<50   

  good_credit  income  income_class  
0         Yes    79.0  Middle Class  
1         Yes    77.0  Middle Class  
2         Yes   720.0   Upper Class  
3         Yes   590.0   Upper Class  
4         Yes   285.0   Upper Class

## Q4.

In [73]:
dummies=pd.get_dummies(df['ethnicity'], dtype='int')

df=pd.concat([df, dummies], axis=1)

In [74]:
df.head()

year    age   race               ethnicity     sex debt_to_income_ratio  \
0  2023  65-74  Black  Not Hispanic or Latino  Female              40%-<50   
1  2023  45-54  Black  Not Hispanic or Latino  Female              40%-<50   
2  2023  55-64  White      Hispanic or Latino    Male              40%-<50   
3  2023  55-64  White  Not Hispanic or Latino  Female              40%-<50   
4  2023  35-44  Asian  Not Hispanic or Latino    Male              40%-<50   

  good_credit  income  income_class  Hispanic or Latino  \
0         Yes    79.0  Middle Class                   0   
1         Yes    77.0  Middle Class                   0   
2         Yes   720.0   Upper Class                   1   
3         Yes   590.0   Upper Class                   0   
4         Yes   285.0   Upper Class                   0   

   Not Hispanic or Latino  
0                       1  
1                       1  
2                       0  
3                       1  
4                       1

In [75]:
df.drop('ethnicity', axis=1, inplace=True)

In [76]:
df.columns

Index(['year', 'age', 'race', 'sex', 'debt_to_income_ratio', 'good_credit',
       'income', 'income_class', 'Hispanic or Latino',
       'Not Hispanic or Latino'],
      dtype='object')

In [77]:
df.rename(columns={'Hispanic or Latino':'ethnicity_Hispanic or Latino', 'Not Hispanic or Latino':'ethnicity_Not Hispanic or Latino'}, inplace=True)

In [78]:
df.head()

year    age   race     sex debt_to_income_ratio good_credit  income  \
0  2023  65-74  Black  Female              40%-<50         Yes    79.0   
1  2023  45-54  Black  Female              40%-<50         Yes    77.0   
2  2023  55-64  White    Male              40%-<50         Yes   720.0   
3  2023  55-64  White  Female              40%-<50         Yes   590.0   
4  2023  35-44  Asian    Male              40%-<50         Yes   285.0   

   income_class  ethnicity_Hispanic or Latino  \
0  Middle Class                             0   
1  Middle Class                             0   
2   Upper Class                             1   
3   Upper Class                             0   
4   Upper Class                             0   

   ethnicity_Not Hispanic or Latino  
0                                 1  
1                                 1  
2                                 0  
3                                 1  
4                                 1

## *Q5*

In [79]:
df.columns

Index(['year', 'age', 'race', 'sex', 'debt_to_income_ratio', 'good_credit',
       'income', 'income_class', 'ethnicity_Hispanic or Latino',
       'ethnicity_Not Hispanic or Latino'],
      dtype='object')

In [80]:
df['income'].mean()

196.65961732124873

In [81]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5958 entries, 0 to 5957
Data columns (total 10 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   year                              5958 non-null   int64  
 1   age                               5958 non-null   object 
 2   race                              5958 non-null   object 
 3   sex                               5958 non-null   object 
 4   debt_to_income_ratio              5958 non-null   object 
 5   good_credit                       5958 non-null   object 
 6   income                            5958 non-null   float64
 7   income_class                      5958 non-null   object 
 8   ethnicity_Hispanic or Latino      5958 non-null   int64  
 9   ethnicity_Not Hispanic or Latino  5958 non-null   int64  
dtypes: float64(1), int64(3), object(6)
memory usage: 465.6+ KB


In [82]:
df['income'].var()

246217.63298706186

In [83]:
np.percentile(df['income'], [5,95])

array([ 49., 473.])

In [84]:
df['income'].median()

130.0

### a.

year, age, race, sex, debt_to_income_ratio, good_credit, income, income_class, ethnicity_Hispanic or Latino, ethnicity_Not Hispanic or Latino

---


### b.

year : Quantitative

age : Quantitative

race : Qualitative - Nominal

sex : Qualitative - Nominal

debt_to_income_ratio : Qualitative - Nominal

good_credit : Qualitative - Nominal

income : Quantitative

income_class : Qualitative - Ordinal

ethnicity_Hispanic or Latino : Qualitative - Ordinal

ethnicity_Not Hispanic or Latino : Qualitative - Ordinal

---

### c.

1. mean : 196.65961732124873

2. variance : 246217.63298706186

3. 5% percentile : 49

4. 95% percentile : 473

5. median : 130
